Source/WebCore/platform/graphics/cocoa/ANGLEUtilitiesCocoa.mm

```diff
@@ -80,6 +121,77 @@ void destroyPbufferAndDetachIOSurface(EGLDisplay display, void* handle)
+        descriptor_spi.skipSampleValidationAndApplySampleAtTileGranularity = YES;
+    descriptor_spi.mutability = MTLMutabilityDefault;
+    descriptor_spi.minFactor  = 0.01;
+
+    auto mtlScreenSize = MTLSizeMake(screenSize.width(), screenSize.height(), 0);
+    if (horizontalSamplesLeft.size() > mtlScreenSize.width || horizontalSamplesRight.size() > mtlScreenSize.width || verticalSamples.size() > mtlScreenSize.height)
+        return nullptr;
+
+    RetainPtr<MTLRasterizationRateLayerDescriptor> layerDescriptorLeft = [[MTLRasterizationRateLayerDescriptor alloc] initWithSampleCount:mtlScreenSize];
+
+    memcpy([layerDescriptorLeft horizontalSampleStorage], horizontalSamplesLeft.data(), horizontalSamplesLeft.size_bytes());
+    memcpy([layerDescriptorLeft verticalSampleStorage], verticalSamples.data(), verticalSamples.size_bytes());
+    [layerDescriptorLeft setSampleCount:MTLSizeMake(horizontalSamplesLeft.size(), verticalSamples.size(), 0)];
+
+    RetainPtr<MTLRasterizationRateLayerDescriptor> layerDescriptorRight = [[MTLRasterizationRateLayerDescriptor alloc] initWithSampleCount:mtlScreenSize];
+
+    memcpy([layerDescriptorRight horizontalSampleStorage], horizontalSamplesRight.data(), horizontalSamplesRight.size_bytes());
+    memcpy([layerDescriptorRight verticalSampleStorage], verticalSamples.data(), verticalSamples.size_bytes());
+    [layerDescriptorRight setSampleCount:MTLSizeMake(horizontalSamplesRight.size(), verticalSamples.size(), 0)];
+
+    [descriptor setScreenSize:mtlScreenSize];
+    [descriptor layers][0] = layerDescriptorLeft.get();
+    [descriptor layers][1] = layerDescriptorRight.get();
+
+#if USE(APPLE_INTERNAL_SDK)
+    auto rateMap = cp_proxy_process_rasterization_rate_map_create(mtlDevice, cp_layer_renderer_layout_shared, 2); // shared, 1);
+    cp_rasterization_rate_map_update_shared_from_layered_descriptor(rateMap, descriptor.get());
+
+    RetainPtr<MTLRasterizationRateMap> rasterizationRateMap = cp_proxy_process_rasterization_rate_map_get_metal_maps(rateMap).firstObject;
+
+    MTLSize physical_size = [rasterizationRateMap physicalSizeForLayer:0];
+    MTLSize physical_size1 = [rasterizationRateMap physicalSizeForLayer:1];
+    auto mtlPhysicalSizeLeft = MTLSizeMake(physicalSizeLeft.width(), physicalSizeLeft.height(), 0);
+    ASSERT_UNUSED(physical_size, physical_size.width == mtlPhysicalSizeLeft.width);
+    ASSERT_UNUSED(mtlPhysicalSizeLeft, physical_size.height == mtlPhysicalSizeLeft.height);
+    auto mtlPhysicalSizeRight = MTLSizeMake(physicalSizeRight.width(), physicalSizeRight.height(), 0);
+    ASSERT_UNUSED(physical_size1, physical_size1.width == mtlPhysicalSizeRight.width);
+    ASSERT_UNUSED(mtlPhysicalSizeRight, physical_size1.height == mtlPhysicalSizeRight.height);
+
+    RELEASE_ASSERT(rasterizationRateMap.get());
+#else
+    UNUSED_PARAM(physicalSizeLeft);
+    UNUSED_PARAM(physicalSizeRight);
+    RetainPtr<MTLRasterizationRateMap> rasterizationRateMap;
+#endif
```

Comment:
This is a mess. I think we should have a separate function for obtaining the rate map from CompositorServices (in a separate file).

Reply:
We could remove lines 167-174 if you think that would help, they are not used other than for assertions.

Reply:
I will remove these assertions, they are not needed